### DISCUSSION TITLE
Mermaid with Icons

### DISCUSSION THREAD

**Discussion by HelgeKrueger** [squidfunk/mkdocs-material#6778] [Q&A]
Consider the simple mermaid diagram:

```
flowchart LR
    A[fa:fa-user Alice] --> B[fa:fa-user Bob]
```

See also https://mermaid.live/edit#pako:eNpNj7EOwjAMRH8l8tz-QAYkKthggQ3SwSRuGymJqzQRQlX_nVCkgifr-e50nkGzIZDQOX7qAWMSp4sKosz-3qHssM4TRbF3VlMr6nonmn_e8KOFCjxFj9aUnPljVpAG8qRAltXZfkgKVFiKEHPi6ytokClmqiCPBhMdLPYRPZRYN230aGziuMERw435J6L1fP7WX79Y3k0NRRw

I want the icons to properly be included in mkdocs-material.

I followed [Additional CSS](https://squidfunk.github.io/mkdocs-material/customization/?h=css#additional-css) to add the fontawesome.css that should contain the style definition, but the icons don't show up.

The relevant mkdocs.yml should be

```yml
extra_css:
  - stylesheets/fontawesome.css
markdown_extensions:
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: mermaid
          format: !!python/name:pymdownx.superfences.fence_code_format
```

**Comment by kamilkrzyskow** [comment]
Hello @HelgeKrueger,
please provide a minimal example reproduction, with all used files. Without it I can only guess blindly. The site you've provided does define the Font Awesome font family and then the font family is set to be used in that Mermaid diagram, and later the CSS for the diagram defines something like:
```css
.fa-user::before {
  content: "\f007";
}
```
which is the code for the "letter" representing that human icon. I don't know what is not working on your end. Perhaps you only need to enforce the usage of the font family in the diagram or perhaps something else is amiss.

> https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/
> https://squidfunk.github.io/mkdocs-material/setup/changing-the-fonts/#customization

**Comment by HelgeKrueger** [comment]
Attached is a minimal failing example:

[example.zip](https://github.com/squidfunk/mkdocs-material/files/14241217/example.zip)

From looking at it from inspector, it seems as if the CSS is not being applied properly. The stylesheet is being loaded though.

**Reply by kamilkrzyskow** [reply]
Bummer, I believe that the closed shadow root prevents external CSS to style the contents. 
So there is nothing you can do without forking the theme and recompiling it, modifying the TypeScript:
https://github.com/squidfunk/mkdocs-material/blob/55603a70e843a8b5e3a4a2ac80c51435c2533600/src/templates/assets/javascripts/components/content/mermaid/index.ts#L118-L124
> https://squidfunk.github.io/mkdocs-material/customization/#building-the-theme

I also have the idea of modifying the `bundle.js` file in the `on_post_build` event in a hook,
The `bundle` is minified and obfuscated, however there is surely a way to find the "mermaid" related contents with Regex and hack the file to your needs. This solution would of course be rather fragile.
> https://www.mkdocs.org/user-guide/configuration/#hooks

**Reply by kamilkrzyskow** [reply]
I wanted to try the hacky method with the hook for myself, and it works @HelgeKrueger 🤩 
However, I wasted a lot of time with that fontawesome.css file, I don't know what's the purpose of it, but it doesn't include the `font-face` definitions... so I had to use the `all.css` variant. I'm using a CDN version because I didn't want to gather all of the fonts and package them.

[mermaid-css.zip](https://github.com/squidfunk/mkdocs-material/files/14251046/mermaid-css.zip)

<details>
<summary> the hook that modifies the bundle.js </summary>

```py
from pathlib import Path


def on_post_build(config):
    site_dir = Path(config["site_dir"])
    javascripts_dir = site_dir / "assets" / "javascripts"

    for path in javascripts_dir.glob("bundle*.js"):
        if str(path).lower().endswith(".js"):
            bundle_path = path
            break
    else:
        bundle_path = None

    if not bundle_path:
        return

    contents = bundle_path.read_text(encoding="utf8")
    # bundle_path.write_text(contents.replace('mode:"closed"', 'mode:"open"'))

    # Find this piece of code in the minimized obfuscated bundle.js
    # /* Create a shadow root and inject diagram */
    # const shadow = host.attachShadow({ mode: "closed" })
    # shadow.innerHTML = svg

    # mode:"closed" has only 1 instance in the bundle.js
    # it only appears for the shadow root in the mermaid file
    mode_pos = contents.find('mode:"closed"')

    # find the first occurrence of the innerHTML after the mode_pos
    innerhtml_pos = contents.find(".innerHTML=", mode_pos + 1)

    # find the first occurrence of the , single line JavaScript separator
    separator_pos = contents.find(",", innerhtml_pos + 1)

    # based on the positions figure out the variable name
    obfuscated_variable = contents[innerhtml_pos:separator_pos].replace(".innerHTML=", "")

    # glue everything together with our custom changes
    # add <style> that imports the custom CSS file with absolute path
    modified_contents = (
        contents[:innerhtml_pos]
        + f".innerHTML=\"<style> @import url('https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css');</style>\"+{obfuscated_variable}"
        + contents[separator_pos:]
    )

    # (optionally) change the mode to open, so that it's possible to modify it later from extra.js
    bundle_path.write_text(modified_contents.replace('mode:"closed"', 'mode:"open"'))
    # bundle_path.write_text(modified_contents)

```

</details>

Notes:
- This approach is rather crazy, but unless the logic of the mermaid changes I think it will work reliably 
- This approach works on assumptions and doesn't have enough safe guards, so once it breaks it will break hard, possibly invalidating the whole bundle.js and breaking search and other bundle.js features
- So good luck :v:

---

A better example, configurable and with safe guards can be found here: 
- https://github.com/squidfunk/mkdocs-material/discussions/6844#discussioncomment-8670472

**Reply by kamilkrzyskow** [reply]
@squidfunk are there any plans to open the shadow root for Mermaid? Or is that hack / recompilation the only solution?

@HelgeKrueger let me know if you want to use the hack in production, I could add some safe guards to prevent total breakage of the site if the hack stops working. 
EDIT: or I could minimize the `bundle.js` changes to only the `closed` -> `open` mode, and then change the styling within `extra.js`, but there would be a delay to loading the styles.

---

A better example, configurable and with safe guards can be found here: 
- https://github.com/squidfunk/mkdocs-material/discussions/6844#discussioncomment-8670472

**Reply by squidfunk** [reply]
I'm not sure we can open the shadow root. We need to put Mermaid diagrams in shadow roots because they will otherwise pollute the `id` space. Other than that, we currently have no plans of opening it, because it might lead to other side effects we are currently not aware of, so very thorough testing would need to be done, which I am currently lacking the time for.

I recommend studying those two issues, which talk a lot about problems with Mermaid diagrams. Some of those problems _might_ be solved in more recent versions of Mermaid.js. If so, we could adapt our solution:

- https://github.com/squidfunk/mkdocs-material/issues/2170
- https://github.com/squidfunk/mkdocs-material/issues/693

**Reply by kamilkrzyskow** [reply]
So in https://github.com/squidfunk/mkdocs-material/issues/2170#issuecomment-882082074 @facelessuser is mentioned as the one who suggested the usage of a shadow root. On his docs, I see that he's using [open shadow roots](https://github.com/facelessuser/pymdown-extensions/blob/87e649dc00fa3d7909d02f0f6109e6eafb712a70/docs/src/js/uml.js#L29) with the Material theme. So I would like to ask him for a second opinion about the "because it might lead to other side effects we are currently not aware of, so very thorough testing would need to be done" claim.

The link from https://github.com/squidfunk/mkdocs-material/issues/2170#issuecomment-770115780 is broken BTW:
- https://github.com/facelessuser/pymdown-extensions/blob/master/docs/src/markdown/_snippets/uml.txt#L5

However, this might be a case of the [X Y Problem](https://xyproblem.info/)

I don't really want to change the shadow root mode, but it was a means to an end of being able to style the insides of the mermaid shadow root with an external JavaScript script, but perhaps this suggestion is unnecessary.

There is an open Issue for mermaid styling:
- https://github.com/squidfunk/mkdocs-material/issues/5619

@squidfunk 
What I would like to propose is to add an option to be able to add custom styles to the mermaid shadow root, like I did in the hook above.
```yaml
extra:
  extra_shadow_root_css:
    - assets/stylesheets/mermaid.css # I don't know if relative paths are supported in the shadow root
    - https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css
```

From my testing I don't see the link to be fetched many times, the browser is smart enough to fetch it only once despite including it in every mermaid block.

**Reply by squidfunk** [reply]
Thanks for suggesting. We have to be very careful for the maintenance nightmare that is adding more and more things under `extra`, and honestly, this looks very, very weird. We would need to somehow pull all of this through from `mkdocs.yml` to the application where we render Mermaid.js. What if we add another integration for another library that also mandates shadow root? Another setting in `extra`? My point is that this feels like a huge bandaid that I'm not comfortable maintaining.

If there are no side effects, we can open the shadow root. However, I need to trust you on that one, because of my currently limited time due to the heavy focus on the big topics to push this project further.

**Reply by kamilkrzyskow** [reply]
Great, I shall wait a bit for [facelessuser](https://github.com/facelessuser) to reply, and later will do a bit more testing with the open mode override in docs that use them, like then one mentioned in the issue above 👍 

EDIT:
I meant `extra_shadow_root_css` as a general extra key for all libraries that would be inside a shadow root. The css can style according to IDs and other unique values for each injected shadow root.

**Reply by squidfunk** [reply]
> I meant extra_shadow_root_css as a general extra key for all libraries that would be inside a shadow root. The css can style according to IDs and other unique values for each injected shadow root.

I know, but sorry, no, we can't introduce more an more degrees of freedom. Overseeing the impact of changes in the light of users being able to customize everything is already a gargantuan task, I'm not interested in making my work even harder 😅

**Reply by matthiasPOE** [reply]
So what is currently the correct way to get the icons working on mermaid charts? Having the icons available especially on a huge chart is a must to distinguish between certain elements

**Reply by mark4409** [reply]
Yes, I would also like to know this

**Reply by squidfunk** [reply]
Please use the search. This question has been asked several times, and IIRC, some workarounds have been proposed.

**Reply by facelessuser** [reply]
I just saw this and realize I was mentioned but never responded:

> So in https://github.com/squidfunk/mkdocs-material/issues/2170#issuecomment-882082074 @facelessuser is mentioned as the one who suggested the usage of a shadow root. On his docs, I see that he's using [open shadow roots](https://github.com/facelessuser/pymdown-extensions/blob/87e649dc00fa3d7909d02f0f6109e6eafb712a70/docs/src/js/uml.js#L29) with the Material theme. So I would like to ask him for a second opinion about the "because it might lead to other side effects we are currently not aware of, so very thorough testing would need to be done" claim.

In the early days of when I added Mermaid support, there were real issues with how SVG elements were managed with Mermaid in general. It was far more reliable to generate Mermaid images on the backend and include them (like they did in their own documentation, though this has changed in recent versions). Some issues where objects disappearing in unrelated diagrams when one was hidden in a Tab. These weren't problems I was guessing at, they were well understood. The shadow root approach fixed many of these issues by isolating things within the shadow root. There was no shared, faulty objects across diagrams. This was the main reason for implementing the shadow root.

> I don't really want to change the shadow root mode, but it was a means to an end of being able to style the insides of the mermaid shadow root with an external JavaScript script, but perhaps this suggestion is unnecessary.

Styling was not really a motivation to use shadow root.

In recent versions of Mermaid, I have found more and more diagrams are viable as many are better formed and more consistent. The main diagram I wouldn't use from the basics is probably the gannt chart which still can exhibit some weird scaling issues at times. This may or may not be due to some combinations of options I use, but it is what it is. My findings are only suggestions.

I stand by the decision to use the shadowroot back when I implemented. The issues I was facing were very well understood, and it was absolutely necessary to work around the inconsistent and poorly managed rendering of earlier Mermaid versions. Particularly bad behaviors with diagrams that were hidden and unhidden when nested in tabs. But it is likely that it could be abandoned moving forward. I have not yet tested some of the problematic cases that caused me to start using shadowroots without shadowroots.

For my uses, I might not abandon the shadowroot as things work great for me as is, and I haven't yet run into a situation where I personally need something that is not compatible with my approach, but that certainly could change if I run into such a scenario. What I do though doesn't have to be what guides Material for MkDocs. I will simply state that the shadowroot, in the most recent versions of Mermaid, may not be necessary, but it was in earlier versions.
